### PR TITLE
[blog] console bin init

### PR DESCRIPTION
**Pull request by TomasVotruba**
*This is a draft! Feel free to send PR with fixes.*

Blog Based on https://github.com/eventigo/eventigo-web/pull/19#issuecomment-301817899 and many personal experiences with not-so-nice console usage in my mentorings.

What do you think @f3l1x ? :dancer:

### CODE REVIEW

File: source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+Only with different `Application` class.
+
+### Protip: use "Application" naming for entry class to your application
+
+As you can see, they are standard way to let user know, where to start. 
+
+
+## Show me Real Refactoring!
+
+If you'd like to see a real life code, [I made PR on Github](htthttps://eventigo.cz/ps://github.com/eventigo/eventigo-web/pull/19/files) to [Eventigo.cz](https://eventigo.cz/) recently.
```
Comment (filipsuk) [inline source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md]: looks like typo in PR link?

Comment (TomasVotruba) [inline source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md]: Yep, I seem to naturally typo every eventigo link :D 

Good catch, thank you.

File: source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+author: 1
+---
+
+`php index.php` is mostly known from [Kdyby\Console](https://github.com/Kdyby/Console/), that integrates [Symfony\Console](https://github.com/symfony/console)
+to Nette DI. Great thanks to [Filip Procházka](https://filip-prochazka.com/) for this package, because it made using console command (something like Presenter, just in CLI)
+very simple. 
+
+## How it Works?
+
+What `php index.php command` actually does?
```
Comment (f3l1x) [inline source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md]: What `php index.php command` actually with `Kdyby\Console` does?

File: source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+```bash
+php artisan
+```
+
+It is only "a file to delegate CLI arguments to". Why not use the same approach in Nette? 
+
+
+## What the Best Practise (so far)
+
+I'm happy to see [this fresh commit](https://github.com/Kdyby/Console/commit/db9c3304f0998bc82724665d3b43d3b6e3eb40ce) by Filip Prochazka, that promotes the same
```
Comment (f3l1x) [inline source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md]: It's all about Kdyby\Console? :-)

File: source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+You may notice it's the same as [`www/index.php`](https://github.com/nette/sandbox/blob/master/www/index.php):
+
+```php
+<?php declare(strict_types=1);
+
+/** @var Nette\DI\Container $container */
+$container = require __DIR__ . '/../app/bootstrap.php';
+
+/** @var Nette\Application\Application $application */
+$application = $container->getByType(Nette\Application\Application::class);
```
Comment (f3l1x) [inline source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md]: I would compare only:

- `$application = $container->getByType(Nette\Application\Application::class);`
- `$consoleApplication = $container->getByType(Symfony\Component\Console\Application::class);`

File: source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+### Protip: use "Application" naming for entry class to your application
+
+As you can see, they are standard way to let user know, where to start. 
+
+
+## Show me Real Refactoring!
+
+If you'd like to see a real life code, [I made PR on Github](https://github.com/eventigo/eventigo-web/pull/19/files) to [Eventigo.cz](https://eventigo.cz/) recently.
+
+We use lightweight and simple [Contributte\Console](https://github.com/contributte/console) by [Milan Šulc](https://jfx.cz/), that also
```
Comment (f3l1x) [inline source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md]: https://f3l1x.io is my new domain ;-)

Comment (f3l1x) [inline source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md]: Milan Felix Šulc is my preferable "nickname".

File: source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+<div class="text-center">
+    <img src="/assets/images/posts/2017/nette-console/commands-after.png">
+</div>
+
+
+## Now You Know
+
+- **The best practise for running console command** in Nette in 2017.
+- **How to refactor** to `bin/console`.
+- There might be some awesome packages in [Contributte](https://github.com/Contributte) :).
```
Comment (f3l1x) [inline source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md]: 👍

**Comment by fprochazka** [comment]
Some notes:

* The **real** historical reason was generating URLs. In Nette 2.1 (or some version close to that), the only way to generate an url was to have a presenter do it. There was no `LinkGenerator`. Therefore you **had to** use `Nette\Application`, have it create a presenter for you and **only then** you were able to generate urls.
* This is stayed in Kdyby becase
    * It is really simple to just install Kdyby/Console and start using it
    * BC
    * I was lazy to update the docs
* It was **always** possible to just have your own `bin/console` and use Symfony Console directly.

File: source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+When you run this command from your CLI, it will:
+ 
+1. use Nette Router
+2. find CliRoute
+3. call Symfony Application (something like [Nette\Application\Application](https://github.com/nette/sandbox/blob/ae3556149309b8442553f6bc70527a923432a19d/www/index.php#L10))
+4. pass it your command name and arguments
+
+## Historical Reasons (Again)
+
+CD players were the best tool to play music on the road... like 10 years ago? Every best practise we use in present has its expiration date.
```
Comment (fprochazka) [inline source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md]: this was never a best practise, but a solution to real-life problem ;)

File: source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md (hunk 3 above)
Comment (fprochazka) [inline source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md]: Is that a problem? :P

Comment (f3l1x) [inline source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md]: Not at all, I was just wondering.. :-)

**Comment by fprochazka** [comment]
To be honest, I didn't like how you [degraded removing Kdyby/Console to removing of a "routing hack"](https://github.com/eventigo/eventigo-web/pull/19) and I feel like the article has the same tone.

File: source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md (hunk 3 above)
Comment (TomasVotruba) [inline source/_posts/2017/2017-05-29-best-practise-for-console-bin-in-nette.md]: I wanted to point a nice event, that we out all now agree on single practise.
Very rare in open-source to see :)

**Comment by TomasVotruba** [comment]
@fprochazka Let's improve it then. How could I express my opinion about redundant code?

**Comment by TomasVotruba** [comment]
@f3l1x Thanks for your comments, I will resolve them after mergin Filip's notes.

**Comment by fprochazka** [comment]
I agree its important to tell people to stop using the routing hack. I also have nothing against promoting Contributte. But lets be fair and not compare apples and oranges.

I'd suggest a simpler, more fair structure for the article

* If we want people to stop using the hack, there should be a short explanation why it was introduced in the first place and that the reason no longer exists. That can be done in few senteces without comparing it to a deprecated best practice.
* Provide a better option - `bin/console`
* A list of relevant integrations of Symfony Console at the end

If you wanna also promote Contributte, a fair comparision would be better than just listing its features without saying the same can be done in Kdyby - it implies that it cannot.

**Comment by fprochazka** [comment]
I agree its important to tell people to stop using the routing hack. I also have nothing against promoting Contributte. But lets be fair and not compare apples and oranges.

IMHO it's important to first clear out what is the article about. Is it about dropping the `php index.php` ? Okey, then I'd suggest a simpler, more fair structure for the article

* If we want people to stop using the hack, there should be a short explanation why it was introduced in the first place and that the reason no longer exists. That can be done in few senteces without comparing it to a deprecated best practice.
* Provide a better option - `bin/console`
* A list of relevant integrations of Symfony Console at the end

If you wanna also promote Contributte, a fair comparision would be better than just listing its features without saying the same can be done in Kdyby - it implies that it cannot.

If you wanna **just** promote Contributte. Do a separate article introducing it.

If you wanna write article about why it's better, then do a fair comparision.

**Comment by fprochazka** [comment]
Also, another note - you can't just tell people to drop using `php index.php` without giving them a replacement. This article is doing (as it is written right now) exactly the same thing you've done in that eventigo PR. You're suggesting to remove the `php index.php` practice without giving a full picture. The article says nothing about `console: url: ...` which is an essential part of the problem (and solution).

**Comment by TomasVotruba** [comment]
@fprochazka Yes, please! Feel free to add the other note as well.

**Comment by fprochazka** [comment]
I'm on it :)

**Comment by TomasVotruba** [comment]
@fprochazka That's great. I'm glad you got your attention on this :+1:

**Comment by fprochazka** [comment]
@TomasVotruba thank you :)

**Comment by fprochazka** [comment]
PR #291

**Comment by TomasVotruba** [comment]
Closing in favor of #291

**Pull request closed without merging**